chips/stm32f4xx/src/chip_specific.rs

```diff
@@ -0,0 +1,269 @@
+    /// Maximum allowed system clock frequency in MHz
+    pub const SYS_CLOCK_FREQUENCY_LIMIT_MHZ: usize = if cfg!(any(
+        feature = "stm32f410",
+        feature = "stm32f411",
+        feature = "stm32f412",
+        feature = "stm32f413",
+        feature = "stm32f423"
+    )) {
+        100
+    } else if cfg!(any(
+        feature = "stm32f405",
+        feature = "stm32f407",
+        feature = "stm32f415",
+        feature = "stm32f417",
+        feature = "stm32f427",
+        feature = "stm32f429",
+        feature = "stm32f437",
+        feature = "stm32f439",
+        feature = "stm32f446",
+        feature = "stm32f469",
+        feature = "stm32f479"
+    )) {
+        // TODO: Some of these models support overdrive model. Change this constant when overdrive support
+        // is added.
+        168
+    } else {
+        //feature = "stm32f401"
+        84
+    };
+}
+
+/// Chip-specific flash code
+pub mod flash_specific {
+    // All this hassle is caused by the fact that the following 4 chip models support 3 bit latency
+    // values, while the other chips support 4 bit values
+    #[cfg(not(any(
+        feature = "stm32f405",
+        feature = "stm32f415",
+        feature = "stm32f407",
+        feature = "stm32f417"
+    )))]
+    #[derive(Copy, Clone, PartialEq, Debug)]
+    /// Enum representing all the possible values for the flash latency
+    pub(crate) enum FlashLatency {
+        /// 0 wait cycles
+        Latency0,
+        /// 1 wait cycle
+        Latency1,
+        /// 2 wait cycles
+        Latency2,
+        /// 3 wait cycles
+        Latency3,
+        /// 4 wait cycles
+        Latency4,
+        /// 5 wait cycles
+        Latency5,
+        /// 6 wait cycles
+        Latency6,
+        /// 7 wait cycles
+        Latency7,
+        /// 8 wait cycles
+        Latency8,
+        /// 9 wait cycles
+        Latency9,
+        /// 10 wait cycles
+        Latency10,
+        /// 11 wait cycles
+        Latency11,
+        /// 12 wait cycles
+        Latency12,
+        /// 13 wait cycles
+        Latency13,
+        /// 14 wait cycles
+        Latency14,
+        /// 15 wait cycles
+        Latency15,
+    }
+
+    #[cfg(any(
+        feature = "stm32f405",
+        feature = "stm32f415",
+        feature = "stm32f407",
+        feature = "stm32f417"
+    ))]
+    #[derive(Copy, Clone, PartialEq, Debug)]
+    /// Enum representing all the possible values for the flash latency
+    pub(crate) enum FlashLatency {
+        /// 0 wait cycles
+        Latency0,
+        /// 1 wait cycle
+        Latency1,
+        /// 2 wait cycles
+        Latency2,
+        /// 3 wait cycles
+        Latency3,
+        /// 4 wait cycles
+        Latency4,
+        /// 5 wait cycles
+        Latency5,
+        /// 6 wait cycles
+        Latency6,
+        /// 7 wait cycles
+        Latency7,
+    }
+
+    // Chip-specific trait that allows to read the current configured flash latency
+    pub(crate) trait SpecificFlashTrait {
```

Comment:
Hmm.. if I understand what's happening correctly, the concept here is that the _only_ file with `#cfg[`'s unrelated to unit testing is this `chip_specific.rs` file.

While I'm always supportive of efforts to contain and ease reasoning about `#define`-hell, in this case I think it makes things a bit more confusing as part of the flash implementation is no longer in the `flash.rs` file.

This is an interesting ergonomics question though. Maybe something we should bring up briefly on a core call.

Reply:
> Hmm.. if I understand what's happening correctly, the concept here is that the _only_ file with `#cfg[`'s unrelated to unit testing is this `chip_specific.rs` file.

Yes, as we did with the earlgrey chip.

> While I'm always supportive of efforts to contain and ease reasoning about `#define`-hell, in this case I think it makes things a bit more confusing as part of the flash implementation is no longer in the `flash.rs` file.

I think we can just leave the config/options/settings here and all the flash stuff can move to flash.rs. The earlgray chip really is a good example: https://github.com/tock/tock/blob/master/chips/earlgrey/src/chip_config.rs

Reply:
So basically leave this as-is, only as a function called from the `flash.rs` rather than a trait on the Flash object?

The big difference I see here is that earlgrey is all const's. Here, there is runtime code difference—critically a different number of frequency ranges to check per chip. This will always* have to be a function somewhere whose body is `cfg`'d.

*I guess you could do something like a array of (mhz_threshold, Latency)'s where the array is the `cfg`'d const and then iterate the array at runtime, but that seems a lot worse and more confusing than what is effectively picking the right jump table currently. Or probably have some macro that creates the right match block, but that again feels really opaque.

Reply:
> So basically leave this as-is, only as a function called from the `flash.rs` rather than a trait on the Flash object?

Yeah

Reply:
Okay, so I looked at this again, decided it wouldn't work because the flash_specific needed to call back into the flash object to read the right register, but then decided that could work, and ultimately just did it to check myself

https://github.com/tock/tock/commit/a9b8c6d2359af8a9d3e96d2c7dca5b92029f7236 — I think this is what @bradjc is getting at. I think this is a bit cleaner, and I like that everything that is specific to a given flash instance is cleanly scoped behind `flash_specific::`

Reply:
I don't like the fact that `chip_specific.rs` depends on `flash.rs`, but if you prefer this solution, I'm happy with it.

Reply:
This is what I am comfortable with: https://github.com/tock/tock/compare/pr/3528?expand=1

With those commits we only reference chips we claim to support and not variants outside of that list. We also only affirmatively use features so it is very clear exactly when certain code is being selected. With any `not()` when that code will get compiled is ambiguous. Also chips which we do not claim to support will not compile until they are explicitly added.

This also removes the conditionally included flash trait and replaces it with chip-specific functions. The flash driver then uses these functions to implement the functionality that was in the trait. This simplifies the code constructs in the chip_specific file to reduce complexity.

Note, I'm not doubting that the extended list of stm variants were correctly mapped to their various settings. But, in general I want to avoid giving the impression of more support than Tock actually offers, and I don't want there to be an example of including code/config that isn't ever used/tested so that others don't copy. If those settings for variants are hard to find then I think it would be fine to include them as comments for future developers who wish to add support for different variants and can then easily include them.

Reply:
The problem with `#[cfg]` for `const` is that clippy tests don't pass anymore. I tried finding a solution for this without success. Any help?

I prefer the idea of commenting out unused features, but I'm also fine with removing them completely. They are not hard to find in documentation.

Personally, I don't like decoupling flash-related functionality from the `Flash` struct, but since you think it's clearer I'll implement it.

Reply:
I added a commit which makes `cargo test` pass for me.

This isn't germane to this issue I guess but the fact that we run into these kind of issues continues to suggest to me that features continue to be a hindrance and should be avoided. The fact that the tests passed before not because of any sort of principled selection of hardware, with intentional testing, but instead because of which branch of the `if` statements happened to be selected is worrying to me.